Documentation/man/man3/HelperFunctions.3
.TH "HelperFunctions" 3 "Fri Mar 17 2017" "Version 1" "targeter" \" -*- nroff -*-
.ad l
.nh
.SH NAME
HelperFunctions
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <HelperFunctions\&.h>\fP
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static cv::Mat \fBdisplayHistogram\fP (int histSize, cv::Mat b_hist, int hist_w=512, int hist_h=400)"
.br
.ti -1c
.RI "static cv::Mat \fBputMatScale\fP (cv::Mat im, bool scale=true, bool bRed=true)"
.br
.ti -1c
.RI "static void \fBresizeImagePreserve\fP (cv::Mat &imageIn, cv::Mat &imageOut, int newWidth, int newHeight)"
.br
.ti -1c
.RI "static void \fBDebugPrintMat\fP (cv::Mat mat)"
.br
.ti -1c
.RI "static cv::Mat \fBtestReturn\fP (cv::Mat &im)"
.br
.ti -1c
.RI "static int \fBreflect\fP (int M, int x)"
.br
.ti -1c
.RI "static std::string \fBtype2str\fP (int type)"
.br
.ti -1c
.RI "static bool \fBisGrayImage\fP (cv::Mat img)"
.br
.ti -1c
.RI "static QImage \fBmakeQImage\fP (cv::Mat im, bool bRGBSwap=true)"
.br
.ti -1c
.RI "static QImage::Format \fBgetFormat\fP (int type)"
.br
.ti -1c
.RI "template<typename T > static cv::Mat \fBputImageScale\fP (T *im, int w, int h)"
.br
.ti -1c
.RI "template<typename T > static cv::Mat \fBputImage\fP (T *im, int width, int height)"
.br
.ti -1c
.RI "template<typename T > static T * \fBgetImage\fP (cv::Mat &m)"
.br
.ti -1c
.RI "template<typename T > static void \fBrescale\fP (T *im, int w, int h)"
.br
.in -1c
.SH "Detailed Description"
.PP 
class for useful helper image manipulation functions 
.PP
Definition at line 11 of file HelperFunctions\&.h\&.
.SH "Member Function Documentation"
.PP 
.SS "void HelperFunctions::DebugPrintMat (cv::Mat mat)\fC [static]\fP"
prints 2D matrix to screen
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/07
.RE
.PP
FullName \fBHelperFunctions::DebugPrintMat\fP Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fIcv::Mat\fP mat 
.RE
.PP
\fBReturns:\fP
.RS 4
void Access public 
.RE
.PP

.PP
Definition at line 342 of file HelperFunctions\&.cpp\&.
.SS "cv::Mat HelperFunctions::displayHistogram (int histSize, cv::Mat b_hist, int hist_w = \fC512\fP, int hist_h = \fC400\fP)\fC [static]\fP"
Returns an OpenCV image from a histogram (b_hist)
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/07
.RE
.PP
FullName \fBHelperFunctions::displayHistogram\fP Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fIint\fP histSize 
.br
\fIcv::Mat\fP b_hist 
.br
\fIint\fP hist_w 
.br
\fIint\fP hist_h 
.RE
.PP
\fBReturns:\fP
.RS 4
cv::Mat Access public 
.RE
.PP
Draw for each channel
.PP
Draw for each channel 
.PP
Definition at line 71 of file HelperFunctions\&.cpp\&.
.SS "QImage::Format HelperFunctions::getFormat (int type)\fC [static]\fP"
Returns QT Image type for OpenCV type
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/07
.RE
.PP
FullName \fBHelperFunctions::getFormat\fP Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fIint\fP type 
.RE
.PP
\fBReturns:\fP
.RS 4
QT_NAMESPACE::QImage::Format Access public 
.RE
.PP

.PP
Definition at line 207 of file HelperFunctions\&.cpp\&.
.SS "template<typename T > static T* HelperFunctions::getImage (cv::Mat & m)\fC [inline]\fP, \fC [static]\fP"
Returns a greyscale 1D int* image array from an OpenCV image
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/07
.RE
.PP
FullName \fBHelperFunctions::getImage\fP Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fIcv::Mat\fP & m 
.RE
.PP
\fBReturns:\fP
.RS 4
T* Access public 
.RE
.PP

.PP
Definition at line 135 of file HelperFunctions\&.h\&.
.SS "bool HelperFunctions::isGrayImage (cv::Mat img)\fC [static]\fP"
determines if image is greyscale
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/07
.RE
.PP
FullName \fBHelperFunctions::isGrayImage\fP Qualifier // returns true if the given 3 channel image is B = G = R 
.PP
\fBParameters:\fP
.RS 4
\fIcv::Mat\fP img 
.RE
.PP
\fBReturns:\fP
.RS 4
bool Access public 
.RE
.PP

.PP
Definition at line 242 of file HelperFunctions\&.cpp\&.
.SS "QImage HelperFunctions::makeQImage (cv::Mat im, bool bRGBSwap = \fCtrue\fP)\fC [static]\fP"
make QImage from OpenCV image
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/07
.RE
.PP
FullName \fBHelperFunctions::makeQImage\fP Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fIcv::Mat\fP im 
.br
\fIbool\fP bRGBSwap 
.RE
.PP
\fBReturns:\fP
.RS 4
QT_NAMESPACE::QImage Access public 
.RE
.PP

.PP
Definition at line 181 of file HelperFunctions\&.cpp\&.
.SS "template<typename T > static cv::Mat HelperFunctions::putImage (T * im, int width, int height)\fC [inline]\fP, \fC [static]\fP"
creates a opencv image from int* 1D image array
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/08
.RE
.PP
FullName putImage Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fIT\fP * im 
.br
\fIint\fP width 
.br
\fIint\fP height 
.RE
.PP
\fBReturns:\fP
.RS 4
cv::Mat Access public static 
.RE
.PP

.PP
Definition at line 104 of file HelperFunctions\&.h\&.
.SS "template<typename T > static cv::Mat HelperFunctions::putImageScale (T * im, int w, int h)\fC [inline]\fP, \fC [static]\fP"
creates a scaled opencv image from int* 1D image array
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/08
.RE
.PP
FullName putImageScale Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fIT\fP * im 
.br
\fIint\fP w 
.br
\fIint\fP h 
.RE
.PP
\fBReturns:\fP
.RS 4
cv::Mat Access public static 
.RE
.PP

.PP
Definition at line 51 of file HelperFunctions\&.h\&.
.SS "cv::Mat HelperFunctions::putMatScale (cv::Mat im, bool scale = \fCtrue\fP, bool bRed = \fCtrue\fP)\fC [static]\fP"
returns a 0-255 scaled opencv image
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/07
.RE
.PP
FullName \fBHelperFunctions::putMatScale\fP Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fIcv::Mat\fP im 
.br
\fIbool\fP scale 
.br
\fIbool\fP bRed 
.RE
.PP
\fBReturns:\fP
.RS 4
cv::Mat Access public 
.RE
.PP

.PP
Definition at line 301 of file HelperFunctions\&.cpp\&.
.SS "int HelperFunctions::reflect (int M, int x)\fC [static]\fP"
Handles vector boundaries by reflecting values
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/07
.RE
.PP
FullName \fBHelperFunctions::reflect\fP Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fIint\fP M 
.br
\fIint\fP x 
.RE
.PP
\fBReturns:\fP
.RS 4
int Access public 
.RE
.PP

.PP
Definition at line 271 of file HelperFunctions\&.cpp\&.
.SS "template<typename T > static void HelperFunctions::rescale (T * im, int w, int h)\fC [inline]\fP, \fC [static]\fP"
rescales image pixel values
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/07
.RE
.PP
FullName \fBHelperFunctions::rescale\fP Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fIT\fP * im 
.br
\fIint\fP w 
.br
\fIint\fP h 
.RE
.PP
\fBReturns:\fP
.RS 4
void Access public 
.RE
.PP

.PP
Definition at line 182 of file HelperFunctions\&.h\&.
.SS "void HelperFunctions::resizeImagePreserve (cv::Mat & imageIn, cv::Mat & imageOut, int newWidth, int newHeight)\fC [static]\fP"
Resizes image while preserving aspect ratio
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/07
.RE
.PP
FullName \fBHelperFunctions::resizeImagePreserve\fP Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fIcv::Mat\fP & imageIn 
.br
\fIcv::Mat\fP & imageOut 
.br
\fIint\fP newWidth 
.br
\fIint\fP newHeight 
.RE
.PP
\fBReturns:\fP
.RS 4
void Access public 
.RE
.PP

.PP
Definition at line 151 of file HelperFunctions\&.cpp\&.
.SS "string HelperFunctions::type2str (int type)\fC [static]\fP"
Returns a string representation of the opencv image type
.PP
\fBAuthor:\fP
.RS 4
David Watts 
.RE
.PP
\fBSince:\fP
.RS 4
2017/03/07
.RE
.PP
FullName \fBHelperFunctions::type2str\fP Qualifier 
.PP
\fBParameters:\fP
.RS 4
\fIint\fP type 
.RE
.PP
\fBReturns:\fP
.RS 4
std::string Access public 
.RE
.PP

.PP
Definition at line 31 of file HelperFunctions\&.cpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for targeter from the source code\&.


HelperFunctions.cpp
#include <iostream>
#include <iomanip>
#include <limits>
#include <QDebug>
#include <vector>
#include <QImage>

#include "opencv2/opencv.hpp"
#include "opencv/highgui.h"
#include "mainwindow.h"

#include "globals.h"
#include "HelperFunctions.h"

using namespace std;


/**
*
*  Returns a string representation of the opencv image type
*
* @author    David Watts
* @since     2017/03/07
* 
* FullName   HelperFunctions::type2str
* Qualifier 
* @param     int type 
* @return    std::string
* Access     public 
*/
string HelperFunctions::type2str(int type)
{
	string r;

	uchar depth = type & CV_MAT_DEPTH_MASK;
	uchar chans = 1 + (type >> CV_CN_SHIFT);

	switch (depth) {
	case CV_8U:  r = "8U"; break;
	case CV_8S:  r = "8S"; break;
	case CV_16U: r = "16U"; break;
	case CV_16S: r = "16S"; break;
	case CV_32S: r = "32S"; break;
	case CV_32F: r = "32F"; break;
	case CV_64F: r = "64F"; break;
	default:     r = "User"; break;
	}

	r += "C";
	r += (chans + '0');

	return r;
}

/**
*
*  Returns an OpenCV image from a histogram (b_hist)
*
* @author    David Watts
* @since     2017/03/07
* 
* FullName   HelperFunctions::displayHistogram
* Qualifier 
* @param     int histSize 
* @param     cv::Mat b_hist 
* @param     int hist_w 
* @param     int hist_h 
* @return    cv::Mat
* Access     public 
*/
cv::Mat HelperFunctions::displayHistogram(int histSize, cv::Mat b_hist, int hist_w, int hist_h)
{
	// Draw the histograms for B, G and R
		
	int fontFace = cv::FONT_HERSHEY_SCRIPT_SIMPLEX;
	int fontScale = 1, thickness=1;
	cv::Scalar red(255, 0, 0);
	cv::Scalar green(0, 255, 0);
	cv::Scalar blue(0, 0, 255);
	int maxheight = 0;
	cv::String text;
	cv::Size textSize;
	int baseline = 0;

	// create image for histogram
	cv::Mat histImage(hist_h, hist_w, CV_8UC3, cv::Scalar(0, 0, 0));
	// bin size in image
	double bin_w = (((double)hist_w) / ((double)histSize));

	std::vector<cv::Point> lines;

	try
	{
		/// Draw for each channel
		for (int i = 0; i < histSize; i+=50)
		{
			text = cv::format("%d", i);

			textSize = cv::getTextSize(text, fontFace, fontScale, 1, &baseline);

			maxheight = max(maxheight, textSize.height);

			putText(histImage, text, cv::Point((int)(bin_w*i), hist_h), fontFace, fontScale, red, 8, true);
		}

		maxheight *= 1.5;

		// Normalize the result to [ 0, histImage.rows ] - i.e. fit in screen
		normalize(b_hist, b_hist, 0, histImage.rows - maxheight, cv::NORM_MINMAX);
			
		/// Draw for each channel
		for (int i = 0; i < histSize; i++)
		{
			lines.push_back(cv::Point((int)(bin_w*i), hist_h - maxheight - cvRound(b_hist.at<float>(i))));
		}

		const cv::Point *pts = (const cv::Point*) cv::Mat(lines).data;
		int npts = cv::Mat(lines).rows;

		polylines(histImage, &pts, &npts, 1,
			false,                  // draw closed contour (i.e. joint end to start)
			cv::Scalar(0, 255, 0),		// colour RGB ordering (here = green)
			1, 					// line thickness
			CV_AA, 0);
	}
	catch (cv::Exception& e)
	{
		const char* err_msg = e.what();
		DBOUT("exception caught: " << err_msg << std::endl);
	}

	return histImage;
}

/**
*
*  Resizes image while preserving aspect ratio
*
* @author    David Watts
* @since     2017/03/07
* 
* FullName   HelperFunctions::resizeImagePreserve
* Qualifier 
* @param     cv::Mat & imageIn 
* @param     cv::Mat & imageOut 
* @param     int newWidth 
* @param     int newHeight 
* @return    void
* Access     public 
*/
void HelperFunctions::resizeImagePreserve(cv::Mat &imageIn, cv::Mat &imageOut, int newWidth, int newHeight)
{
	int w = newWidth;
	int h = newHeight;

	double ratio = ((double)imageIn.cols) / ((double)imageIn.rows); // width/height

	int scaledH = (int)round(((double)w) / ratio);
	int scaledW = (int)round(((double)h)*ratio);

	if (scaledH > h) { w = scaledW; }
	else { h = scaledH; }

	cv::resize(imageIn, imageOut, cv::Size(w, h), 0, 0, cv::INTER_LINEAR);
}

/**
*
*  make QImage from OpenCV image
*
* @author    David Watts
* @since     2017/03/07
*
* FullName   HelperFunctions::makeQImage
* Qualifier
* @param     cv::Mat im
* @param     bool bRGBSwap
* @return    QT_NAMESPACE::QImage
* Access     public
*/
QImage HelperFunctions::makeQImage(cv::Mat im, bool bRGBSwap)
{
	QImage::Format form = getFormat(im.type());

	QImage pImage = QImage(im.data, im.cols, im.rows, (int)im.step, form);

	if (bRGBSwap)
		pImage.rgbSwapped();

	return pImage;
}


/**
*
*  Returns QT Image type for OpenCV type
*
* @author    David Watts
* @since     2017/03/07
*
* FullName   HelperFunctions::getFormat
* Qualifier
* @param     int type
* @return    QT_NAMESPACE::QImage::Format
* Access     public
*/
QImage::Format HelperFunctions::getFormat(int type)
{

#ifdef DEBUGPRINT
	qDebug() << "Function Name: " << Q_FUNC_INFO;
#endif

	switch (type) {
	case CV_8UC3: {
		return QImage::Format_RGB888;
	}
	case CV_8U: {
		return QImage::Format_Indexed8;
	}
	case CV_8UC4: {
		return  QImage::Format_ARGB32;
	}
				  return QImage::Format_Invalid;
	}
	return QImage::Format_Invalid;
}

/**
*
*  determines if image is greyscale
*
* @author    David Watts
* @since     2017/03/07
* 
* FullName   HelperFunctions::isGrayImage
* Qualifier  // returns true if the given 3 channel image is B = G = R
* @param     cv::Mat img 
* @return    bool
* Access     public 
*/
bool HelperFunctions::isGrayImage(cv::Mat img) // returns true if the given 3 channel image is B = G = R
{
	for (int j = 0; j < img.rows; j++)
	{
		for (int i = 0; i < img.cols; i++)
		{
			if (img.ptr<uchar>(j)[(i * 3)] != img.ptr<uchar>(j)[(i * 3) + 1] ||
				(img.ptr<uchar>(j)[(i * 3)] != img.ptr<uchar>(j)[(i * 3) + 2]))
				return false;
		}
	}
	return true;
}


/**
*
*  Handles vector boundaries by reflecting values
*
* @author    David Watts
* @since     2017/03/07
* 
* FullName   HelperFunctions::reflect
* Qualifier 
* @param     int M 
* @param     int x 
* @return    int
* Access     public 
*/
int HelperFunctions::reflect(int M, int x)
{
	if (x < 0)
	{
		return -x - 1;
	}

	if (x >= M)
	{
		return 2 * M - x - 1;
	}

	return x;
}

/**
*
*  returns a 0-255 scaled opencv image
*
* @author    David Watts
* @since     2017/03/07
* 
* FullName   HelperFunctions::putMatScale
* Qualifier 
* @param     cv::Mat im 
* @param     bool scale 
* @param     bool bRed 
* @return    cv::Mat
* Access     public 
*/
cv::Mat HelperFunctions::putMatScale(cv::Mat im, bool scale, bool bRed)
{
	cv::Mat cMat;

	cv::normalize(im, cMat, 255, 0);

	return cMat;
}


cv::Mat HelperFunctions::testReturn(cv::Mat& im)
{
	cv::Mat cMat = cv::Mat(im.rows, im.cols, CV_8UC3, cv::Scalar(0));

	for (int j = 0; j < im.rows; j++)
	{
		for (int i = 0; i < im.cols; i++)
		{
			cMat.ptr<uchar>(j)[(i * 3)] = 0;
			cMat.ptr<uchar>(j)[(i * 3) + 1] = j % 255;
			cMat.ptr<uchar>(j)[(i * 3) + 2] = i % 255;
		}
	}

	return cMat;
}


/**
*
*  prints 2D matrix to screen
*
* @author    David Watts
* @since     2017/03/07
* 
* FullName   HelperFunctions::DebugPrintMat
* Qualifier 
* @param     cv::Mat mat 
* @return    void
* Access     public 
*/
void HelperFunctions::DebugPrintMat(cv::Mat mat)
{
	for (int i = 0; i < mat.cols; i++)
	{
		for (int j = 0; j < mat.rows; j++)
		{
			std::cout << (int)mat.at<uchar>(j, i) << " ";
		}
	}
}

